### ISSUE TITLE
[feature] A way to query values in settings.yml from conanfile.py

### ISSUE THREAD

**Issue by samuel-emrys** [conan-io/conan#13793]
### What is your suggestion?

Hi Team,

It would be really useful from a compatibility perspective to be able to query the possible values for a given setting in settings.yml from a conanfile.py. To illustrate this with a usecase, I've been running into issues with libcxx compatibility in tool_requires packages, specifically doxygen as documented in https://github.com/conan-io/conan-center-index/issues/17034 with a potential fix in https://github.com/conan-io/conan-center-index/pull/17051.

Ultimately, this has lead to a `compatibility()` definition as follows:

```python
    @property
    def _conan_home(self):
        conan_home_env = "CONAN_USER_HOME" if Version(conan_version).major < 2 else "CONAN_HOME"
        conan_home_dir = ".conan" if Version(conan_version).major < 2 else ".conan2"
        return os.environ.get(conan_home_env, pathlib.Path(pathlib.Path.home(), conan_home_dir))

    def compatibility(self):
        # is there a better way of reading all possible versions from settings.yml?
        settings_yml_path = pathlib.Path(self._conan_home, "settings.yml")
        with open(settings_yml_path, "r") as f:
            settings_yml = yaml.safe_load(f)

        compatible_versions = [{"settings": [("compiler.version", v), ("build_type", "Release")]}
            for v in settings_yml["compiler"][str(self.settings.compiler)]["version"] if v <= Version(self.settings.compiler.version)]
        return compatible_versions
```

This is necessary, because this compatibility is based on possible compiler versions - if the available binary was built with an older version of the compiler then it will be compatible, so my understanding is that I need to provide a list of the possible values an "older version of the compiler" could possibly have here. 

This raises additional questions about the way libcxx is being modelled (similar to the questions around libc), where the compiler is being used as a heuristic to model the libcxx dependency, but these are probably out of scope for this issue so lets leave that for another thread.

The above `compatibility()` definition works when creating/consuming the package locally for me, but alas it does not pass the sniff test when executing `conan info` on the CI servers because it fails to locate `settings.yml`. I'm not sure how robust my logic is above to identifying the conan home, but it strikes me that it would be useful to have a direct hook into the conan home being used by the conanfile at the time of interpretation, rather than relying on some imperfect heuristic as I am above. This would also be useful for custom user settings to model the libc compatibility described in https://github.com/conan-io/docs/issues/3960.

If there's a better way to approach this, any advice is appreciated.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Thanks for the suggestion @samuel-emrys 

I agree, this would be useful and necessary. We already have an internal ``get_definition()`` for this, it would be just re-thinking the UI if necessary and documenting it to make it public.

**Comment by samuel-emrys** [comment]
Thanks @memsharded this is a good steer. Looks like the following works well in conan 2:

```python
    def compatibility(self):
        compatible_versions = [{"settings": [("compiler.version", v), ("build_type", "Release")]}
            for v in self.settings.compiler.version.get_definition() if v <= Version(self.settings.compiler.version)]
        return compatible_versions
```
It's not the most intuitive name, but it's exactly what I was looking for. All that's missing is documentation.

Interestingly though, the above seems to get the `settings.compiler.version` object for the build profile, rather than the host profile - is that intentional?

**Comment by memsharded** [comment]
I am adding the following test, it seems to work fine:
```python
def test_settings_definitions_compiler():
    c = TestClient()
    conanfile = textwrap.dedent("""
        from conan import ConanFile
        class Pkg(ConanFile):
            settings = "compiler"
            def generate(self):
                definition = self.settings.compiler.version.get_definition()
                self.output.info("HOST: " + ", ".join(definition))
                definition = self.settings_build.compiler.version.get_definition()
                self.output.info("BUILD: " + ", ".join(definition))
            """)
    c.save({"conanfile.py": conanfile})
    # New settings are there
    c.run("install . -s compiler=msvc -s compiler.version=192 -s:b compiler=gcc")
    assert "conanfile.py: HOST: 170, 180, 190, 191, 192" in c.out
    assert "conanfile.py: BUILD: 4.1, 4.4, 4.5, 4.6, 4.7," in c.out
```
I will submit the test in a PR to also discuss possible naming for ``get_definition()``, any suggestion?

**Comment by samuel-emrys** [comment]
Ah, thanks for that. The issue I was having was when building https://github.com/conan-io/conan-center-index/pull/17051 with a gcc host profile and clang build profile. It would build the package, but because it's a `tool_requires`, the test_package tests it as a `tool_requires`, so the `test_package` would try to find a package matching the build profile. Mystery solved!

Thinking about the naming convention, I've struggled to identify anything better, but I'm also trying to understand why I didn't make the association earlier given that it doesn't seem to have an easy replacement. I think perhaps partially because the way that I've thought about the settings object is as a single coordinate in the cloud of possible settings values - not the cloud itself, probably predominantly because this is mostly how the `self.settings` object is currently used in recipes. `get_definitions()` returns the cloud of possible values.

I think I would have found it a lot more intuitive if it was `self.settings.get_definition()["compiler.version"]` because this would mean the definition is being obtained over all settings, but I'm not convinced that this is a _better_ interface, even though subjectively I find it more intuitive.

Honestly, it's not perfect but it's probably pretty reasonable as is. Just adding some documentation would go a long way to improving comprehension.

**Comment by memsharded** [comment]
> I think I would have found it a lot more intuitive if it was self.settings.get_definition()["compiler.version"]

See https://github.com/conan-io/conan/pull/13796/files, it seems ``self.settings.get_definition()["compiler"]["gcc"]["version"]`` works, which sounds good.

**Comment by memsharded** [comment]
This was finally named ``.possible_values()``, will be released in 2.0.5
